kubeadm: apply "master" label/taint migration for 1.24

neolit123 opened the pull request:
#### What type of PR is this?

/kind cleanup feature

#### What this PR does / why we need it:

adapt the kubeadm managed labels and taints to changes planned in release 1.24, as part of the effort to remove usages of the word "master".

short summary:
- for new clusters, CP nodes will only have the new "control-plane" label ("master" label is no longer applied).
- for new clusters, CP nodes will be tainted with both the "master" and "control-plane" taint
- for upgraded clusters, "kubeadm upgrade apply" will remove the "master" label from CP nodes.
- for upgraded clusters, "kubeadm upgrade apply" will apply "control-plane" taint (next to the existing "master" taint). the "master" taint will be removed in 1.25.

details from the KEP:
https://git.k8s.io/enhancements/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint

1.24 = stage 2

```
## Design Details

The process will be broken into multiple stages:
- First - 1.20
- Second - Minimum deprecation period for GA features is 1 year.
Estimated 1.24, but may depend on user feedback.
- Third - one release after Second
- Fourth - one release after Third

### Renaming the "node-role.kubernetes.io/master" Node label

First stage:
- Introduce the "node-role.kubernetes.io/control-plane" label in parallel to
the "master" label.
- Announce to users that they should adapt to use the new label.

Second stage:
- Remove the "master" label and announce it to the users.

### Renaming the "node-role.kubernetes.io/master" Node taint

First stage:
- Introduce the "node-role.kubernetes.io/control-plane:NoSchedule" toleration
in the CoreDNS Deployment of kubeadm.
- Announce to users that they should do that same for their workloads.

Second stage:
- Add the "node-role.kubernetes.io/control-plane:NoSchedule" taint to Nodes.

Third stage:
- Remove the "node-role.kubernetes.io/master:NoSchedule" taint from Nodes.

Fourth stage:
- Remove the "node-role.kubernetes.io/master:NoSchedule" toleration in the CoreDNS
Deployment of kubeadm
- Announce to users that they should remove tolerations for the "master" taint in
their workloads.
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubeadm/issues/2200

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
ACTION REQUIRED: kubeadm: apply "second stage" of the plan to migrate kubeadm away from the usage of the word "master" in labels and taints. For new clusters, the label "node-role.kubernetes.io/master" will no longer be added to control plane nodes, only the label "node-role.kubernetes.io/control-plane" will be added. For clusters that are being upgraded to 1.24 with "kubeadm upgrade apply", the command will remove the label "node-role.kubernetes.io/master" from existing control plane nodes. For new clusters, both the old taint "node-role.kubernetes.io/master:NoSchedule" and new taint "node-role.kubernetes.io/control-plane:NoSchedule" will be added to control plane nodes. In release 1.20 ("first stage"), a release note instructed to preemptively tolerate the new taint. For clusters that are being upgraded to 1.24 with "kubeadm upgrade apply", the command will add the new taint "node-role.kubernetes.io/control-plane:NoSchedule" to existing control plane nodes. Please adapt your infrastructure to these changes. In 1.25 the old taint "node-role.kubernetes.io/master:NoSchedule" will be removed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://git.k8s.io/enhancements/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint
```

neolit123 committed:
kubeadm: delete the old "master" label during upgrade

- Rename the function in postupgrade.go to better reflect
what is being done.
- During "upgrade apply" find all nodes with the old label
and remove it by calling PatchNode.
- Update health check for CP nodes to not track "master"
labeled nodes. At this point all CP nodes should have
"control-plane" and we can use that selector only.

neolit123 committed:
kubeadm: only apply the new "control-plane" label during init/join

- Update the markcontrolplane phase used by init and join to
only label the nodes with the new control plane label.
- Cleanup TODOs about the old label.
- Remove outdated comment about selfhosting in staticpod/utils.go.
Selfhosting has not been supported in kubeadm for a while
and the comment also mentions the "master" label.
- Update unit tests.

neolit123 committed:
kubeadm: apply the new "control-plane" taint on CP nodes

- Apply "control-plane" taint during init/join by adding the
taint in SetNodeRegistrationDynamicDefaults(). The old
taint "master" is still applied.
- Clarify API docs (v1beta2 and v1beta3) for nodeRegistration.Taint
to not mention "master" taint and be more generic. Remove
example for taints that includes the word "master".
- Update unit tests.

neolit123 committed:
kubeadm: apply the new "control-plane" taint during upgrade

- During "upgrade apply" call a new function AddNewControlPlaneTaint()
that finds all nodes with the new "control-plane" node-role label
and adds the new "control-plane" taint to them.
- The function is called in "apply" and is separate from
the step to remove the old "master" label for better debugging
if errors occur.

neolit123 commented:
/triage accepted
/priority important-soon
/milestone v1.24

neolit123 commented:
/approve cancel

neolit123 commented:
in theory the kind jobs should pass with this change.
AFAIK, the kind test clusters have 2 untainted worker nodes, which the test suite will find and use instead of the tainted CP nodes.

```
Run 'kubectl get nodes' on the control-plane to see this node join the cluster.
 ✓ Joining worker nodes 🚜
 • Waiting ≤ 1m0s for control-plane = Ready ⏳  ...
 ✗ Waiting ≤ 1m0s for control-plane = Ready ⏳
 • WARNING: Timed out waiting for Ready ⚠️
Set kubectl context to "kind-kind"
You can now use your cluster with:
```

looks like the CP node did not get status `Ready` in 1m, so i wonder if it never got `Ready` after that.
i will do more testing next week.

neolit123 commented:
i think the problem is the `master` label removal:
https://github.com/kubernetes-sigs/kind/blob/b6bc112522651d98c81823df56b7afa511459a3b/pkg/cluster/internal/create/actions/waitforready/waitforready.go#L91

kind uses the `master` label as selector for node ready checks, and we have to change it for `>= 1.24`.

neolit123 committed:
test/e2e_kubeadm: update label/taint e2e tests for 1.24

- Reword old "master" label to "Legacy"
- Add new "control-plane" label
- Use new label to count CP nodes
- Check for both taints on CP nodes
- Leave TODOs for 1.25

neolit123 commented:
/retest

neolit123 commented:
@pacoxu as you pointed out here:
> pull-kubernetes-e2e-kind can set non-blocking-taints with flag. Or we should change its default value?

i guess we might have to update:
https://github.com/kubernetes/kubernetes/blob/e045e5c875b2f51c4c4b1bc14d4e717123649b17/test/e2e/framework/test_context.go#L319
to tolerate `control-plane` and `master` taints at the same time for this release.

the logic in the test framework seems non-optimal to me:
https://github.com/kubernetes/kubernetes/blob/6296c6e4d336f9d3e4cc80e0af5d4bc74f330d0d/test/e2e/framework/node/wait.go#L261-L278

the kind cluster has 3 nodes, 1 is the CP node which has 2 unchedulable taints and is Ready and 2 worker nodes without taints that are also ready. so even if there are sufficient nodes to run tests on (2 are the minimum allowed for tests) the framework will block on the CP node that has one unknown taint...

pacoxu commented:
> to tolerate `control-plane` and `master` taints at the same time for this release.

OK. According to your comments in https://github.com/kubernetes-sigs/kind/pull/2595#discussion_r788825554, we will tolerate both in 1.24 and remove `master` in 1.25 or later.

BenTheElder commented:
there's some very naive and outdated assumptions in e2e.test + cluster/ about "master" nodes, I recall at least one other place looking by node name previously, there are probably more we haven't spotted here yet : /

pretty up there with the conflation of --provider with very specific cluster bringup tools

neolit123 committed:
e2e/framework/test_context: add "control-plane" to non-blocking-taints

Having only the "master" taint in the list of non-blocking taints
blocks kubeadm / kind clusters from migrating to applying
both the "control-plane" and "master" taints in 1.24.
Add "control-plane" to the list of taints.

Leave TODO to cleanup the "master" taint in 1.25+.
It has to be removed either way as part of the inclusive
language cleanup efforts.

neolit123 commented:
> OK. According to your comments in kubernetes-sigs/kind#2595 (comment), we will tolerate both in 1.24 and remove master in 1.25 or later.

yes, i've updated the PR to have the "control-plane" taint in the non-blocking-taints default list. also left a TODO to cleanup the "mater" taint in 1.25+. i suspect it might break users that don't use k8s originated tools like kOps / kubeadm.

> there's some very naive and outdated assumptions in e2e.test + cluster/ about "master" nodes, I recall at least one other place looking by node name previously, there are probably more we haven't spotted here yet : /

i think `@oomichi` managed to clean up the node name binding. looks like the "master" taint is only used in one device plugin YAML (sig-node) but other than that and test/e2e_kubeadm i don't see usage.

i did see a number of `master` words in the framework though and it seems after WG naming disbanded the work to clean it up has stalled.

neolit123 commented:
/assign @SataQiu 
PTAL for approval.

neolit123 commented:
after some discussion with release team, this is not tracked as a KEP for 1.24, but the KEP from 1.20 still applies.
/milestone clear

SataQiu reviewed (approved):
/approve

found a small typo in release note:
`For cluster that are` -> `For clusters that are`

neolit123 commented:
> found a small typo in release note:

thanks, fixed.

looking for LGTM on this, @SataQiu @pacoxu

SataQiu reviewed (commented):
/lgtm

pacoxu commented:
/retest 
for flake

pacoxu commented:
/retest
for flake
TestReflectorNotStopedOnSlowInitialization
